### ISSUE TITLE
Add a location/geoip tag in device attributes

### ISSUE THREAD

**Issue by nicola-lunghi** [homieiot/convention#78]
as per title

**Comment by ingoogni** [comment]
IMO that should be part of your mqtt string if geolocation is important, in principle it is nothing different from "front/door" or "headlight/left".

**Comment by timpur** [comment]
I agree with @ingoogni, the only thought i had was if we had device stats implemented as a node, then this is something you could add to that stat node to make your own device stats. So would be nice to also open up device stats for custom user device stats i guess? @nicola-lunghi, would this satisfy this?

**Comment by ThomDietrich** [comment]
When last skimming the issue I had a similar question. Which kind of location are we talking about?

* "1st floor, left bedroom"
* [geo:51.6848,15.9205](https://en.wikipedia.org/wiki/Geo_URI_scheme)
* ...

**Comment by nicola-lunghi** [comment]
maybe I can implement this using a property node could be more flexible

**Comment by timpur** [comment]
The reason i was thinking a device stat is because who said locations were fixed? The location could change... device $attributes are not meant to change over time? or has that changed also?

**Comment by ingoogni** [comment]
If geolocation is important, for a moving device, set up a node for it. I think a geo location spec would belong in $unit and would fit the Convention well. (Next question I can imagine is yaw, roll, pitch ..... all just a node with '$unit'.)

**Comment by euphi** [comment]
A location string is very useful for auto-discovery, so I opt for adding it to the convention as optional field with arbitrary string, e.g. a room name (For example openhab automatically groups devices by location). 

This location usually don't change often - if you move a device to another room, you can change the configuration of the device. 

If you a moving device, e.g. a drone, the geolocation would be a normal node, of course.

**Comment by nicola-lunghi** [comment]
in my setup i need something like an x/y coordinate (geographical
coordinate are not very practical in a building)

**Comment by ingoogni** [comment]
"A location string is very useful for auto-discovery, so I opt for adding it to the convention as optional device property with arbitrary string, e.g. a room name"

Lets have a single device with 4 temp probes built in the ceiling on a strategic spot so it can measure the temperature in 4 different rooms, that could even be on a different floor. Is location really a device property? IMHO not.

It can be nice that software groups certain data on location, but it should do it on (semantic) topic strings instead of yet more properties of a device that in most cases can (and should) be expressed as node properties.

**Comment by timpur** [comment]
I'm happy with both sides tbh, so I'll let you guys work this out :P

**Comment by davidgraeff** [comment]
@ingoogni A device is usually an entity that has a single location. But I do understand your concern, that a node might have a different location than the device itself. I would allow both. The location is optional anyway.

From a controller perspective, the device location is of more interest though. A node is usually rendered as part of a device and a different location can not always be considered.

**Comment by ingoogni** [comment]
I have a 'device' that lifts and lowers a hose, it is a spindle with a steppermotor controlled by a Nano over I2C connected to a Wemos over MQTT connected to a Wemos that is the dedicated remote. What is the device? how many devices are there? Now there are also a few end-switches the hard stops connected to the nano, the soft stops connected to the Wemos. There is a hand switch that can be operated from three places, there s a thermo sensor etc. How many devices?

For messaging mqtt topics, for me only one, the all live in the same namespace. From a device point of view, two esps? But why stop counting there, a switch is also a device and yes, in this case its location is important, it is the end switch on the top side.

Maybe namespace is a better word than semantic topic. An ESP with one temp sensor services two namespaces, the sensors one and the ESP one. Wether a node on a mesh is a carrier of other node does not matter and imo should not be seen as something different, they are all nodes that communicate, mqtt, 1-wire, i2c etc.

So, yes, a device constisting of devices consisting of devices etc. can have a location as can have all the sub(sub) assemblies.

**Comment by ingoogni** [comment]
Maybe I should add that what y'all see as device, homie on an ESP to me is just a router in a network that transmits messages from and to endpoint, sensors, actuators or other procedures that happen to run on the ESP.
Now of coarse I have a peek in the logs every now and then to see how the router performs, but it is not the interesting thing in the network, my NAS and local web server are of more interest as that is what I work with, knowing that without the router I could not.

**Comment by ThomDietrich** [comment]
That seems too far fetched. We have to insert an abstraction layer somewhere and a tree in the form homie->devices->nodes->properties should be able to cover most applications at least sufficiently.

Regarding location: I get why a location might be useful for nodes (Heck, one of my projects publishes plant sensor data for multiple Bluetooth-connected devices distributed in the house under the namespace of one device).

As for the convention: I think an optional location for the device should be more important. Additionally I don't see why we shouldn't allow it for nodes as well.

Could we go back to the data format? From what I got, three formats should be considered. In the following I propose three distinguishable schemes in one string:

Topic: `homie/mydevice[/mynode]/$location`
 * An informal location name (room): "1st floor, left bedroom"
* A standardized lat-lon location (geo): "[geo:51.6848,15.9205](https://en.wikipedia.org/wiki/Geo_URI_scheme)"
* A location in a cartesian coordinate system (relative): "(1, 2, 3.456)"

**Comment by ThomDietrich** [comment]
Not sure what you mean with "remove as much boilerplate" as your proposal is more complex.

Regarding geo: fine with me of course

Regarding type: Probably follows our existing scheme better. I don't particularly like that type is an attribute-attribute but it's not a deal breaker.

**Comment by ThomDietrich** [comment]
Topic: `homie/mydevice[/mynode]/$type`
* "text": A non-interpreted string
* "geo": Follow the scheme https://en.wikipedia.org/wiki/Geo_URI_scheme - The "geo:" in the beginning is *optional*
* "relative": cartesian coordinates in meters

Topic: `homie/mydevice[/mynode]/$location`
 * **text**: "1st floor, left bedroom"
* **geo**: "51.6848,15.9205"
* **relative**: "(1, 2, 3.456)"

**Comment by ingoogni** [comment]
Sorry guys, I don't get it, why "invent" $location &type geo: etc when we already have $unit, $datatype and $format in place? Make those available system wide.

And in most cases a fitting topic string will suffice:
a standard example of topicstrings:
http://www.steves-internet-guide.com/understanding-mqtt-topics/
even better:
http://tinkerman.cat/mqtt-topic-naming-convention/

**Comment by ThomDietrich** [comment]
thanks for those links.... here is one for you https://github.com/homieiot/convention#topic-ids

The answer to your question is simple: because $unit etc are for properties of nodes. $location is an attribute itself. From the convention:

> Attributes: Devices, nodes and properties have specific attributes characterizing them. Attributes are represented by topic identifier starting with $. The precise definition of attributes is important for the automatic discovery of devices following the Homie convention.

**Comment by fermuch** [comment]
> * A standardized lat-lon location (geo): "geo:51.6848,15.9205"

@ThomDietrich for latlng pairs, you'd also need a projection (most likely [ESRI 4326](http://spatialreference.org/ref/epsg/wgs-84/)). A projection is a way to plot coordinates on a flat surface (a map). To my knowledge, there are several thousands, and some are required to be used when working for governments.

If we add a $location property, what projection should be used needs to be specified on Homie when using the geo type. Noting decimals precission is also important.

The easiest standard format to encode location for embedded systems, to my knowledge, would be [WKT](https://en.wikipedia.org/wiki/Well-known_text#Geometric_objects) POINT (e.g.: `POINT (-13.7040324, -69.6703397)`), or by using a custom format like the one you specified (I preffer this solution), but both the projection expected and decimal precission need to be noted on the Homie standard if we go for a custom format.

**Comment by davidgraeff** [comment]
@ingoogni Please try to keep your answers short or move the important parts to the beginning and elaborate later on in your post. A "I thing this should be an optional node and I don't like it" would have been sufficient :)

@fermuch The projection is not required for the GEO URI schema. 

@ThomDietrich 
What about `homie/device/location` to make it a normal node and specify "$type=location". Properties are "geo", "text", "relative". So all of them can be given, or just one.

**Comment by fermuch** [comment]
> @fermuch The projection is not required for the GEO URI schema.

Very true! WGS-84 is the default coordinate reference according to the [RFC](https://tools.ietf.org/html/rfc5870) (section 3.4.2).

**Comment by ThomDietrich** [comment]
All, please try to keep your answers short and technical.
In my last summary posting I posted 5 questions. It would be great if you provided your feedback on them.

@fermuch did you follow the link I provided? Here it is once again: https://en.wikipedia.org/wiki/Geo_URI_scheme

@davidgraeff "location" as a normal node/property would be contradictory to the convention as it is right now. The convention only specifies attributes, while it leaves nodes/properties to the application. Attributes are for metadata and discovery, while non-`$` topics (device, node, property, `set`) are for the actual application.
I would prefer to not break with that.

**Comment by davidgraeff** [comment]
$location/geo, $location/relative, $location/text like with the $stats topic sounds like an option.

**Comment by fermuch** [comment]
> @fermuch did you follow the link I provided? Here it is once again: https://en.wikipedia.org/wiki/Geo_URI_scheme

Yeah, that's how I found the RFC. It's on the very first sentence of the wiki.
I'm okay with using geo uris. I was just concerned about forcing everyone to use one kind of projection when requirements might be different for someone else.

After re-reading what I wrote, I might be talking a little bit pedantic, which isn't my intention, it's just that I already suffered a lot with projections and decimal precission in other projects, I don't desire it for anyone else.

**Comment by euphi** [comment]
@ingoogni An IOT device is not an DNS server. (Relative) Location does matter, there are use cases, so it does belong to HomieZero (optional). And because there are also use-cases for the location of the sensors, the location may also belong to a sensor.

BTW, regarding your train example: It is common practice for devices on trains to provide their relative location in device meta-data. There is even a train specific standard [EN 15380-3](https://www.beuth.de/en/standard/din-en-15380-3/84830436) for that...

**Comment by ingoogni** [comment]
The train is a nice one. Do they provide their location like a beacon? If so, how different is that to the random number generator? Do they communicate to establish their relative position?
IMO 'geolocation' is not the core business of Homie, just like timestamps are not the core business of Homie even though both are important aspects of IoT and both deserve require proper specs for $unit, $format etc.

**Comment by ThomDietrich** [comment]
This is derailing. Could we please go back to https://github.com/homieiot/convention/issues/78#issuecomment-385978216 ?

**Comment by euphi** [comment]
@ingoogni geolocation isn't, but location is.

So, to keep things simple I propose that the location is an optional device attribute of name `$location` and type "String". 

I don't think there is a need to add more rules for type to the convention. If someone want's to use geolocation, he can write the "geo:..." uri to the location string. We may add a example for location with geo-string.

**Comment by davidgraeff** [comment]
@euphi You are not thinking about the controller perspective :/ On the controller side, I'm of course interested in the semantic of $location, otherwise it's useless.

**Comment by timpur** [comment]
I kinda agree with @euphi, if you want to use the https://en.wikipedia.org/wiki/Geo_URI_scheme use `geo:` else its just a plain string location like `beach house`. So on the discovery side if you see  start with`geo:` then parse it, else show as is ?

**Comment by davidgraeff** [comment]
Would be a good thing to actually evaluate  topic hierarchy Vs string parsing. So we are back at Thomas proposal, that was quite decent.

**Comment by ThomDietrich** [comment]
Thanks David.
I'd be ready to create a PR. I was simply waiting for responses regarding questions 1, 2, 3, 4 and 5 😅

**Comment by euphi** [comment]
>1. $location/$type or $location-type or something else?

I opt for `$location/$type`. 

> 2. "text"?

`text`is ok and should be the default if  `$type` is ommited.

> 3. "geo: optional?

I opt for arbitrary types. We could propose some useful "official" types (like geo:), but there can be many other possible types, like "room number"  or domain-specific location designations (as the mentioned EN 15380). 

> 4. relative in meters?

Is there a standard for relative coordinates? Wha

> 5. For devices and nodes alike?

I opt for yes, nodes shall also have an optional $location.

**Comment by ThomDietrich** [comment]
I'll take care of this one in the PR following after https://github.com/homieiot/convention/issues/129#issuecomment-442458794

**Comment by mjcumming** [comment]
@ThomDietrich did you ever flesh this out in a PR?
